GPU/D3D11/FramebufferManagerD3D11.cpp

```diff
@@ -686,85 +701,38 @@ void FramebufferManagerD3D11::PackFramebufferD3D11_(VirtualFramebuffer *vfb, int
 	// We always need to convert from the framebuffer native format.
 	// Right now that's always 8888.
 	DEBUG_LOG(HLE, "Reading framebuffer to mem, fb_address = %08x", fb_address);
+	ID3D11Texture2D *colorTex = (ID3D11Texture2D *)draw_->GetFramebufferAPITexture(vfb->fbo, Draw::FB_COLOR_BIT, 0);
 
-	/*
-	LPDIRECT3DSURFACE9 renderTarget = (LPDIRECT3DSURFACE9)draw_->GetFramebufferAPITexture(vfb->fbo, Draw::FB_COLOR_BIT | Draw::FB_SURFACE_BIT, 0);
-	D3DSURFACE_DESC desc;
-	renderTarget->GetDesc(&desc);
-
-	LPDIRECT3DSURFACE9 offscreen = GetOffscreenSurface(renderTarget, vfb);
-	if (offscreen) {
-		HRESULT hr = pD3Ddevice->GetRenderTargetData(renderTarget, offscreen);
-		if (SUCCEEDED(hr)) {
-			D3DLOCKED_RECT locked;
-			u32 widthFactor = vfb->renderWidth / vfb->bufferWidth;
-			u32 heightFactor = vfb->renderHeight / vfb->bufferHeight;
-			RECT rect = { (LONG)(x * widthFactor), (LONG)(y * heightFactor), (LONG)((x + w) * widthFactor), (LONG)((y + h) * heightFactor) };
-			hr = offscreen->LockRect(&locked, &rect, D3DLOCK_READONLY);
-			if (SUCCEEDED(hr)) {
-				// TODO: Handle the other formats?  We don't currently create them, I think.
-				const int dstByteOffset = (y * vfb->fb_stride + x) * dstBpp;
-				// Pixel size always 4 here because we always request BGRA8888.
-				ConvertFromRGBA8888(Memory::GetPointer(fb_address + dstByteOffset), (u8 *)locked.pBits, vfb->fb_stride, locked.Pitch / 4, w, h, vfb->format);
-				offscreen->UnlockRect();
-			} else {
-				ERROR_LOG_REPORT(G3D, "Unable to lock rect from %08x: %d,%d %dx%d of %dx%d", fb_address, rect.left, rect.top, rect.right, rect.bottom, vfb->renderWidth, vfb->renderHeight);
-			}
-		} else {
-			ERROR_LOG_REPORT(G3D, "Unable to download render target data from %08x", fb_address);
-		}
+	D3D11_BOX srcBox{ 0, 0, 0, vfb->width, vfb->height, 1 };
+	context_->CopySubresourceRegion(packTexture_, 0, 0, 0, 0, colorTex, 0, &srcBox);
+
+	// Ideally, we'd round robin between two packTexture_, and simply use the other one. Though if the game
+	// does a once-off copy, that won't work at all.
+
+	// BIG GPU STALL
+	D3D11_MAPPED_SUBRESOURCE map;
+	HRESULT result = context_->Map(packTexture_, 0, D3D11_MAP_READ, 0, &map);
+	if (FAILED(result)) {
+		return;
 	}
-	*/
+
+	// TODO: Handle the other formats?  We don't currently create them, I think.
+	const int srcByteOffset = y * map.RowPitch + x * 4;
+	const int dstByteOffset = (y * vfb->fb_stride + x) * dstBpp;
+	// Pixel size always 4 here because we always request BGRA8888.
+	ConvertFromRGBA8888(Memory::GetPointer(fb_address + dstByteOffset), (u8 *)map.pData, vfb->fb_stride, map.RowPitch/4, w, h, vfb->format);
+	context_->Unmap(packTexture_, 0);
 }
 
+// Nobody calls this yet.
```

Comment:
Right. There are potential uses, some silly games read the depth buffer using the CPU for lens flares, so it would be useful for that as well, though with a heavy speed hit (but we could double buffer it making lens flares one frame late).